parity-util-mem/src/impls.rs

```diff
@@ -73,3 +65,57 @@ impl<T: MallocSizeOf> MallocSizeOf for RwLock<T> {
 		self.read().size_of(ops)
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use crate::{allocators::new_malloc_size_ops, MallocSizeOf, MallocSizeOfOps};
+	use smallvec::SmallVec;
+	impl_smallvec!(3);
+
+	#[test]
+	fn test_smallvec_stack_allocated_type() {
+		let mut v: SmallVec<[u8; 3]> = SmallVec::new();
+		let mut ops = new_malloc_size_ops();
+		assert_eq!(v.size_of(&mut ops), 0);
+		v.push(1);
+		v.push(2);
+		v.push(3);
+		assert_eq!(v.size_of(&mut ops), 0);
+		assert!(!v.spilled());
+		v.push(4);
+		assert!(v.spilled(), "SmallVec spills when going beyond the capacity of the inner backing array");
+		assert_eq!(v.size_of(&mut ops), 4); // 4 u8s on the heap
+	}
+
+	#[test]
+	fn test_smallvec_boxed_stack_allocated_type() {
+		let mut v: SmallVec<[Box<u8>; 3]> = SmallVec::new();
+		let mut ops = new_malloc_size_ops();
+		assert_eq!(v.size_of(&mut ops), 0);
+		v.push(Box::new(1u8));
+		v.push(Box::new(2u8));
+		v.push(Box::new(3u8));
+		assert!(v.size_of(&mut ops) >= 3);
+		assert!(!v.spilled());
+		v.push(Box::new(4u8));
+		assert!(v.spilled(), "SmallVec spills when going beyond the capacity of the inner backing array");
+		let mut ops = new_malloc_size_ops();
+		assert!(v.size_of(&mut ops) >= 36);
+	}
+
+	#[test]
+	fn test_smallvec_heap_allocated_type() {
+		let mut v: SmallVec<[String; 3]> = SmallVec::new();
+		let mut ops = new_malloc_size_ops();
+		assert_eq!(v.size_of(&mut ops), 0);
+		v.push("COW".into());
+		v.push("PIG".into());
+		v.push("DUCK".into());
+		assert!(!v.spilled());
+		assert!(v.size_of(&mut ops) >= 10);
+		v.push("ÖWL".into());
+		assert!(v.spilled());
+		let mut ops = new_malloc_size_ops();
+		assert!(v.size_of(&mut ops) >= 110);
```

Comment:
can we extract these magic constants into something like
`std::mem::size_of::<String>() * 4 + "ÖWL".len() + "COW".len() + "PIG".len() + "DUCK".len()`?